gateway/sds_gateway/visualizations/processing/waterfall.py

```diff
@@ -12,43 +12,173 @@
 from digital_rf import DigitalRFReader
 from django.conf import settings
 from loguru import logger
+from pydantic import BaseModel
+from pydantic import Field
+from pydantic import computed_field
+from pydantic import field_validator
+from pydantic import model_validator
 
 from sds_gateway.api_methods.utils.minio_client import get_minio_client
 
 
-@dataclass
-class WaterfallSliceParams:
-    """Parameters for processing a waterfall slice."""
-
-    reader: Any
-    channel: str
-    slice_idx: int
-    start_sample: int
-    samples_per_slice: int
-    end_sample: int
-    fft_size: int
-    min_frequency: float
-    max_frequency: float
-    sample_rate: float
-    center_freq: float
+class WaterfallSliceParams(BaseModel):
+    """Parameters for processing a waterfall slice with validation."""
+
+    reader: Any = Field(exclude=True)  # Exclude from serialization
+    channel: str = Field(min_length=1, description="Channel name")
+    slice_idx: int = Field(ge=0, description="Slice index")
+    start_sample: int = Field(ge=0, description="Start sample index")
+    samples_per_slice: int = Field(gt=0, description="Number of samples per slice")
+    end_sample: int = Field(gt=0, description="End sample index")
+    fft_size: int = Field(gt=0, description="FFT size")
+    center_freq: float = Field(description="Center frequency in Hz")
+    sample_rate_numerator: int = Field(gt=0, description="Sample rate numerator")
+    sample_rate_denominator: int = Field(gt=0, description="Sample rate denominator")
+
```

Comment:
I think there's nothing specific to waterfall about this dataclass; See if you can repurpose this to also cover `generate_spectrogram_from_drf`, which does a lot of the same data validation.

Reply:
Implemented in https://github.com/spectrumx/sds-code/pull/196